src/fmllauncher/java/net/minecraftforge/fml/loading/toposort/TopologicalSort.java

```diff
@@ -33,21 +37,98 @@
 import java.util.List;
 import java.util.Map;
 import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.PriorityQueue;
+import java.util.Queue;
 import java.util.Set;
 import java.util.SortedSet;
 import java.util.TreeSet;
 import java.util.function.Function;
 import java.util.stream.Collectors;
 
 /**
- * Topological sort for mod loading
- *
- * Based on a variety of sources, including http://keithschwarz.com/interesting/code/?dir=topological-sort
- * @author cpw
+ * Provides a topological sort algorithm.
  *
+ * <p>While this algorithm is used for mod loading in forge, it can be
+ * utilized in other fashions, e.g. topology-based registry loading, prioritization
+ * for renderers, and even mod module loading.
  */
-public class TopologicalSort
-{
+public final class TopologicalSort {
+
+    /**
+     * A breath-first-search based topological sort.
+     *
+     * <p>Compared to the depth-first-search version, it does not reverse the graph
+     * and supports custom secondary ordering specified by a comparator. It also utilizes the
+     * recently introduced Guava Graph API, which is more straightforward than the old directed
+     * graph.
+     *
+     * <p>The graph to sort must be directed, must not allow self loops, and must not contain
+     * cycles. {@link IllegalArgumentException} will be thrown otherwise.
+     *
+     * <p>When {@code null} is used for the comparator and multiple nodes have no
+     * prerequisites, the order depends on the iteration order of the set returned by the
+     * {@link Graph#successors(Object)} call, which is random by default.
+     *
+     * <p>The overall performance is {@code O(E)}, where {@code E} indicates the number of
+     * edges. However, if the comparator performance is worse than {@code O(1)}, then the
+     * overall performance would degrade to {@code O(N * P)}, where {@code N} is the number
```

Comment:
Are those complexities correct? Here's what I'd think would be correct: `O(E+V)` would be more accurate for the case of `comparator==null` (relevant e.g. if there aren't any edges), for the other case I'm pretty sure you're missing a factor of `log V` or even `V` depending on how `PriorityQueue` is implemented (the docs unfortunately don't give any complexity). Insertion/deletion takes `log n` steps/comparisons in a binary heap with n elements (insertion could be improved to `O(1)` using impractical structures), and you're doing `O(n)` insertions and deletions.

Reply:
yeah, should be `O(E + V log(V))` instead. The degraded performance should be `O(E + PV log(V))` as well.